db/db_impl/db_impl_open.cc

```diff
@@ -1158,266 +1171,470 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+                               std::move(file_reader), reporter,
+                               true /*checksum*/, wal_number));
+  return status;
+}
+
+Status DBImpl::ProcessLogRecord(
+    Slice record, const std::unique_ptr<log::Reader>& reader,
+    const UnorderedMap<uint32_t, size_t>& running_ts_sz, uint64_t wal_number,
+    const std::string& fname, bool read_only, int job_id,
+    std::function<void()> logFileDropped, DBOpenLogReporter* reporter,
+    uint64_t* record_checksum, SequenceNumber* next_sequence,
+    bool* stop_replay_for_corruption, Status* status,
+    bool* stop_replay_by_wal_filter,
+    std::unordered_map<int, VersionEdit>* version_edits, bool* flushed) {
+  assert(reporter);
+  assert(stop_replay_for_corruption);
+  assert(status);
+  assert(stop_replay_by_wal_filter);
+
+  Status process_status;
+  bool has_valid_writes = false;
+  WriteBatch batch;
+  std::unique_ptr<WriteBatch> new_batch;
+  WriteBatch* batch_to_use = nullptr;
+
+  if (record.size() < WriteBatchInternal::kHeader) {
+    reporter->Corruption(record.size(),
+                         Status::Corruption("log record too small"));
+    assert(process_status.ok());
+    return process_status;
+  }
+
+  process_status = InitializeWriteBatchForLogRecord(
+      record, reader, running_ts_sz, &batch, new_batch, batch_to_use,
+      record_checksum);
+  if (!process_status.ok()) {
+    return process_status;
+  }
+  assert(batch_to_use);
+
+  SequenceNumber sequence = WriteBatchInternal::Sequence(batch_to_use);
+  if (sequence > kMaxSequenceNumber) {
+    reporter->Corruption(
+        record.size(),
+        Status::Corruption("sequence " + std::to_string(sequence) +
+                           " is too large"));
+    assert(process_status.ok());
+    return process_status;
+  }
+
+  if (immutable_db_options_.wal_recovery_mode ==
+      WALRecoveryMode::kPointInTimeRecovery) {
+    MaybeReviseStopReplayForCorruption(sequence, next_sequence,
+                                       stop_replay_for_corruption);
+    if (*stop_replay_for_corruption) {
+      logFileDropped();
+      assert(process_status.ok());
+      return process_status;
+    }
+  }
+
+  // For the default case of wal_filter == nullptr, always performs no-op
+  // and returns true.
+  if (!InvokeWalFilterIfNeededOnWalRecord(wal_number, fname, *reporter, *status,
+                                          *stop_replay_by_wal_filter,
+                                          *batch_to_use)) {
+    assert(process_status.ok());
+    return process_status;
+  } else {
+    // FIXME(hx235): Handle the potential non-okay `status` when
+    // `InvokeWalFilterIfNeededOnWalRecord()` returns true
+    status->PermitUncheckedError();
+  }
+
+  assert(process_status.ok());
+  process_status = InsertLogRecordToMemtable(batch_to_use, wal_number,
+                                             next_sequence, &has_valid_writes);
+  MaybeIgnoreError(&process_status);
```

Comment:
One of the non-ok status that insert records to memtable can return could be because there is some in memory corruption detected, the checksum doesn't match the initial checksum calculated when data is read in. And here in `!paranoid_check` mode, we ignore such an error.

While during `InitializeWriteBatchForLogRecord`, if `UpdateProtectionInfo` returns non-OK status, which also showcase a in memory corruption, we fail the whole recovery without checking this `MaybeIgnoreError` logic. There are some inconsistencies in how we treat these errors. That could be an item we revisit to make it consistent.

The non-OK status returned by `HandleWriteBatchTimestampSizeDifference` is probably treated with  unnecessary criticality that should as well be treated by the different recovery modes' logic instead of fail the whole thing regardless (this one is added by me  😅).

Reply:
Tracked